contracts/random/PRNG.sol

```diff
@@ -38,31 +41,32 @@ library PRNG {
     }
 
     /**
-    @dev Hashes seed||counter, placing it in the entropy word, and resets the
-    remaining bits to 256. Increments the counter BEFORE the refill (ie 0 is
-    never used) as this simplifies round-tripping with store() and loadSource()
-    because the stored counter state is the same as the one used for deriving
-    the entropy pool.
+    @dev Computes the next PRN in entropy using a lag-1 multiply-with-carry
+    algorithm and resets the remaining bits to 128.
+    `nextNumber = (factor * number + carry) mod 2**128`
+    `nextCarry  = (factor * number + carry) //  2**128`
      */
     function _refill(Source src) private pure {
         assembly {
-            let ctr := add(src, COUNTER)
-            mstore(ctr, add(1, mload(ctr)))
-            mstore(add(src, ENTROPY), keccak256(src, 0x40))
-            mstore(add(src, REMAIN), 256)
+            let carryAndNumber := mload(src)
+            let rand := and(carryAndNumber, 0xffffffffffffffffffffffffffffffff)
+            let carry := shr(128, carryAndNumber)
+            let tmp := add(mul(MWC_FACTOR, rand), carry)
+            mstore(add(src, REMAIN), 128)
+            mstore(src, tmp)
```

Comment:
(nit) I don't think there's a need for `tmp`, and the store into `src` should follow immediately from the logic before, setting remain as a final step. Although minor, both of these reduce cognitive load on the reader ever so slightly because there's no need to remember what `tmp` is nor hold it while figuring out the `REMAIN` logic.